File: dot-product-of-two-sparse-vectors/dot-product-of-two-sparse-vectors.go
type SparseVector struct {
    vector []int
}
​
func Constructor(nums []int) SparseVector {
    return SparseVector{nums}
}
​
// Return the dotProduct of two sparse vectors
func (this *SparseVector) dotProduct(vec SparseVector) int {
    sum := 0
    for i := range vec.vector {
        sum += vec.vector[i] * (*this).vector[i]
    }
    return sum
}
​
/**
 * Your SparseVector object will be instantiated and called as such:
 * v1 := Constructor(nums1);
 * v2 := Constructor(nums2);
 * ans := v1.dotProduct(v2);
 */

File: dot-product-of-two-sparse-vectors/dot-product-of-two-sparse-vectors.py
class SparseVector:
    def __init__(self, nums: List[int]):
        self.dic = {}
        for i in range(len(nums)):
            if nums[i]:
                self.dic[i] = nums[i]

    # Return the dotProduct of two sparse vectors
    def dotProduct(self, vec: 'SparseVector') -> int:
        total = 0
        for i in self.dic:
            if i in vec.dic:
                total += self.dic[i] * vec.dic[i]
        return total

# Your SparseVector object will be instantiated and called as such:
# v1 = SparseVector(nums1)
# v2 = SparseVector(nums2)
# ans = v1.dotProduct(v2)